File: src/uptodown.py
import json
import logging 

from src import session 
from bs4 import BeautifulSoup

def get_latest_version(app_name: str) -> str:
    conf_file_path = f'./apps/uptodown/{app_name}.json'
    with open(conf_file_path, 'r') as json_file:
        config = json.load(json_file)

    url = f"https://{config['name']}.en.uptodown.com/android/versions"

    response = session.get(url)
    response.raise_for_status()
    content_size = len(response.content)
    logging.info(f"URL:{response.url} [{content_size}/{content_size}] -> \"-\" [1]")
    soup = BeautifulSoup(response.content, "html.parser")
    version_spans = soup.select('#versions-items-list .version')
    versions = [span.text for span in version_spans]
    highest_version = max(versions)
    
    return highest_version

def get_download_link(version: str, app_name: str) -> str:
    # Load configuration
    with open(f'./apps/uptodown/{app_name}.json', 'r') as file:
        config = json.load(file)

    base_url = f"https://{config['name']}.en.uptodown.com/android"
    response = session.get(f"{base_url}/versions")
    response.raise_for_status()
    
    soup = BeautifulSoup(response.content, "html.parser")
    data_code = soup.find('h1', id='detail-app-name')['data-code']

    page = 1
    while True:
        response = session.get(f"{base_url}/apps/{data_code}/versions/{page}")
        response.raise_for_status()
        version_data = response.json().get('data', [])
        
        for entry in version_data:
            if entry["version"] == version:
                version_url = entry["versionURL"]
                version_page = session.get(version_url)
                version_page.raise_for_status()
                soup = BeautifulSoup(version_page.content, "html.parser")
                
                # Check for button type
                button = soup.find('button', id='detail-download-button')
                if "download-link-deeplink" in button['onclick']:
                    # Update versionURL by adding '-x'
                    version_url += '-x'
                    version_page = session.get(version_url)
                    version_page.raise_for_status()
                    soup = BeautifulSoup(version_page.content, "html.parser")
                    button = soup.find('button', id='detail-download-button')
                
                download_url = button['data-url']
                return f"https://dw.uptodown.com/dwn/{download_url}"
        
        if all(entry["version"] < version for entry in version_data):
            break
        page += 1

    return None

File: lib/uptodown.js
const https = require("https");
const fs = require("fs");
const path = require("path");

const BASE = "https://youtube.en.uptodown.com/android";

// Helper fetch nội bộ
function fetch(url, headers = {}, isBinary = false) {
  return new Promise((resolve, reject) => {
    https.get(
      url,
      {
        headers: {
          "user-agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/122 Safari/537.36",
          ...headers,
        },
      },
      (res) => {
        let data = [];
        res.on("data", (c) => data.push(c));
        res.on("end", () => {
          const buffer = Buffer.concat(data);
          resolve({
            status: res.statusCode,
            headers: res.headers,
            body: isBinary ? buffer : buffer.toString(),
          });
        });
      }
    ).on("error", reject);
  });
}

async function downloadFromUptodown(version) {
  try {
    console.log("🌐 BASE:", BASE);
    console.log("🔎 VERSION:", version);

    // 1. Lấy appCode
    console.log("⬇️ Fetch versions page...");
    const html = (await fetch(`${BASE}/versions`)).body;

    const codeMatch = html.match(
      /id="detail-app-name"[^>]*data-code="(\d+)"/
    );
    if (!codeMatch) throw new Error("No appCode found");

    const appCode = codeMatch[1];
    console.log("🧩 APP CODE:", appCode);

    // 2. Tìm versionId
    let page = 1;
    let versionId = null;

    while (page < 10) {
      const apiUrl = `${BASE}/apps/${appCode}/versions/${page}`;
      console.log("🌐 PAGE:", apiUrl);

      const res = await fetch(apiUrl);
      const json = JSON.parse(res.body);

      if (!json.data || json.data.length === 0) break;

      const target = json.data.find(
        (item) =>
          item.version === version && item.kindFile === "apk"
      );

      if (target) {
        versionId = target.versionURL.versionID;
        break;
      }

      page++;
    }

    if (!versionId) throw new Error("No version found");

    console.log("➡️ VERSION ID:", versionId);

    // 3. Lấy token
    const pageUrl = `${BASE}/download/${versionId}`;
    console.log("🌐 DOWNLOAD PAGE:", pageUrl);

    const pageRes = await fetch(pageUrl, {
      referer: `${BASE}/versions`,
    });

    const tokenMatch = pageRes.body.match(
      /id="detail-download-button"[^>]*data-url="([^"]+)"/
    );

    if (!tokenMatch) throw new Error("No download token");

    const token = tokenMatch[1];
    const finalUrl = `https://dw.uptodown.com/dwn/${token}`;

    console.log("🔗 FINAL:", finalUrl);

    // 4. Tải file
    const outDir = path.resolve(__dirname, "..", "downloads");
    if (!fs.existsSync(outDir))
      fs.mkdirSync(outDir, { recursive: true });

    const filePath = path.join(
      outDir,
      `youtube-${version.replace(/\./g, "-")}-uptodown.apk`
    );

    console.log("⬇️ Downloading via HTTP...");

    let res = await fetch(finalUrl, { referer: pageUrl }, true);

    // redirect nếu có
    if (
      res.status >= 300 &&
      res.status < 400 &&
      res.headers.location
    ) {
      console.log("🔁 Redirect:", res.headers.location);
      res = await fetch(res.headers.location, {}, true);
    }

    if (res.status === 200) {
      fs.writeFileSync(filePath, res.body);
      console.log("📦 DONE:", filePath);
      return filePath;
    }

    throw new Error("Download failed: " + res.status);
  } catch (err) {
    console.error("❌ ERROR:", err.message);
    throw err;
  }
}

module.exports = { downloadFromUptodown };
